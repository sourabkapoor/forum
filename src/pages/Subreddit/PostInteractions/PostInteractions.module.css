.post-excerpt__interactions {
  display: flex;
  gap: 5px;
  padding-left: 4px;
  padding-right: 8px;
  color: #818384;
  margin-bottom: 5px;
}

.post-excerpt__interaction {
  padding: 8px;
  display: flex;
  align-items: center;
  margin-right: 6px;
  border-radius: 10px;
  cursor: pointer;
  background-color: transparent;
  border: none;
  color: var(--secondary-color);
}

.post-excerpt__interaction:hover {
  /* background-color: rgba(215,218,220,0.1); */
  background-color: rgba(199, 199, 199, 0.443);
  outline: none;
  padding: 8px;
  word-break: normal;
}

.post-excerpt__icon {
  width: 16px;
  height: 16px;
  margin-right: 6px;
}

.post-excerpt__action {
  font-size: 12px;
  white-space: nowrap;
  overflow: hidden;
}


