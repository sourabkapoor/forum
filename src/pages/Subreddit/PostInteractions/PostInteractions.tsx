import CSSModules from "react-css-modules";
import { BiMessage } from "react-icons/bi";
import styles from "./PostInteractions.module.css";
import { IoMdShareAlt } from "react-icons/io";
import { FaRegBookmark } from "react-icons/fa";
import React, { useEffect, useState } from "react";
import { useAppDispatch, useAppSelector } from "../../../hooks/hooks";
import {
  selectAuthStatus,
  toggleSignInModal,
} from "../../../features/auth/authSlice";
import {
  doc,
  setDoc,
  serverTimestamp,
  collection,
  addDoc,
  where,
  query,
  getDoc,
  getDocs,
  deleteDoc,
} from "firebase/firestore";
import { nanoid } from "nanoid";
import { db, getUser, getUserId } from "../../../firebase";
import ModeCommentOutlinedIcon from '@mui/icons-material/ModeCommentOutlined';
import ReplyOutlinedIcon from '@mui/icons-material/ReplyOutlined';

interface Props {
  commentsQuantity: number;
  postId: string | undefined;
}

const PostInteractions: React.FC<Props> = ({ commentsQuantity, postId }) => {
  const isLoggedIn = useAppSelector(selectAuthStatus);
  const dispatch = useAppDispatch();

  const [saveButtonText, setSaveButtonText] = useState("Save");

  async function savePost() {
    if (!isLoggedIn) {
      dispatch(toggleSignInModal());
      return;
    }

    try {
      const docId = `${getUserId()}${postId}`;
      const savedPostsRef = doc(db, "savedPosts", docId);
      console.log('??? ', docId, postId, getUserId())
      await setDoc(savedPostsRef, {
        id: docId,
        postId,
        savedAt: serverTimestamp(),
        userId: getUserId(),
      });
      setSaveButtonText("Saved");
    } catch (error) {
      console.log(`Could not save post: ${error}`);
    }
  }

  async function unSavePost() {
    if (!isLoggedIn) {
      dispatch(toggleSignInModal());
      return;
    }

    try {
      const savedPostDocRef = doc(db, "savedPosts", `${getUserId()}${postId}`);

      await deleteDoc(savedPostDocRef);

      setSaveButtonText("Save");
    } catch (error) {
      console.log(`Could not save posts: ${error}`);
    }
  }

  useEffect(() => {
    async function fetchSaveStatus() {
      if (!isLoggedIn || !postId) {
        return;
      }

      try {
        const savedPostDocRef = doc(
          db,
          "savedPosts",
          `${getUserId()}${postId}`
        );

        const savedPostDoc = await getDoc(savedPostDocRef);

        if (savedPostDoc.exists()) {
          setSaveButtonText("Saved");
        }
      } catch (error) {
        console.log(`Could not fetch post save status: ${error}`);
      }
    }
    fetchSaveStatus();
  }, [postId, dispatch, isLoggedIn]);
  return (
    <div
      styleName="post-excerpt__interactions"
      onClick={(e) => e.preventDefault()}
    >
      <button styleName="post-excerpt__interaction">
        {/* <BiMessage styleName="post-excerpt__icon" /> */}
        <ModeCommentOutlinedIcon styleName="post-excerpt__icon" />
        <span styleName="post-excerpt__action">
          {commentsQuantity} Comments
        </span>
      </button>
      <button styleName="post-excerpt__interaction">
        <ReplyOutlinedIcon styleName="post-excerpt__icon" />
        <span styleName="post-excerpt__action">Share</span>
      </button>
      <button
        styleName="post-excerpt__interaction"
        onClick={saveButtonText === "Save" ? savePost : unSavePost}
      >
        <FaRegBookmark styleName="post-excerpt__icon" />
        <span styleName="post-excerpt__action">{saveButtonText}</span>
      </button>
    </div>
  );
};

export default CSSModules(PostInteractions, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
