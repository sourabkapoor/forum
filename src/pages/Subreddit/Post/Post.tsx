import styles from "./Post.module.css";
import { BiUpvote } from "react-icons/bi";
import { BiDownvote } from "react-icons/bi";
import { BiMessage } from "react-icons/bi";
import Votes from "../Votes/Votes";
import PostAuthor from "../PostAuthor/PostAuthor";
import PostInteractions from "../PostInteractions/PostInteractions";
import {
  addDoc,
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  onSnapshot,
  query,
  QuerySnapshot,
  runTransaction,
  serverTimestamp,
  setDoc,
  where,
  writeBatch,
} from "firebase/firestore";
import CSSModules from "react-css-modules";
import { render } from "@testing-library/react";
import { db, getUserId } from "../../../firebase";
import React, { useEffect, useReducer, useState } from "react";
import { useLocation, useParams } from "react-router-dom";
import { useDispatch } from "react-redux";
import {
  selectAuthStatus,
  toggleSignInModal,
} from "../../../features/auth/authSlice";
import { nanoid } from "nanoid";
import { useAppSelector } from "../../../hooks/hooks";
import { transcode } from "buffer";
import { selectCommunityData } from "../../../features/subreddit/subredditSlice";

interface Props {
  data: DocumentData | undefined;
}

const Post: React.FC<Props> = (props) => {
  let { postId } = useParams();
  if(postId === undefined)
    postId = props?.data?.id 

  const [postData, setPostData] = useState<DocumentData | undefined>();
  const [userPostVote, setUserPostVote] = useState<DocumentData | undefined>();

  const isLoggedIn = useAppSelector(selectAuthStatus);
  const dispatch = useDispatch();
  const location = useLocation();
  const { id: subredditId } = useAppSelector(selectCommunityData);

  useEffect(() => {
    async function fetchPost() {
      const postDocRef = collection(db, "posts");

      const q = query(postDocRef, where("id", "==", postId));

      const postDoc = await getDocs(q);

      // setPostData(postDoc?.docs[0]?.data());
      onSnapshot(q, (snapshot: QuerySnapshot) => {
        console.log(snapshot.docs);
        setPostData({ ...snapshot.docs[0].data(), id: snapshot.docs[0].id });
      });
    }

    props.data ?? fetchPost();
  }, [postId, props.data]);

  const onVote = async (
    vote: number,
    event: React.MouseEvent<HTMLImageElement, MouseEvent>
  ) => {
    event.preventDefault();
    if (!isLoggedIn) {
      dispatch(toggleSignInModal());
      return;
    }

    try {
      await runTransaction(db, async (transaction) => {
        const userPostVotesRef = doc(
          db,
          "users",
          `${getUserId()}/postVotes/${postId ?? props.data?.id}`
        );

        const postRef = doc(db, "posts", postId ?? props.data?.id);

        const post = await transaction.get(postRef);

        const userPostVotes = await transaction.get(userPostVotesRef);
        let voteChange = vote;

        if (!userPostVotes.exists()) {
          const newVote = {
            id: userPostVotesRef.id,
            postId: postId ?? props.data?.id,
            subredditId,
            voteValue: vote,
          };

          transaction.set(userPostVotesRef, newVote);
        } else {
          if (userPostVotes.data().voteValue === vote) {
            voteChange *= -1;
            transaction.update(postRef, {
              voteStatus: post?.data()?.voteStatus - vote,
            });
            transaction.delete(userPostVotesRef);
          } else {
            voteChange = 2 * vote;
            transaction.update(postRef, {
              voteStatus: post?.data()?.voteStatus + 2 * vote,
            });
            transaction.update(userPostVotesRef, {
              voteValue: vote,
            });
          }
        }
        transaction.update(postRef, {
          voteStatus: post?.data()?.voteStatus + voteChange,
        });
      });
    } catch (error) {
      alert(`Could not vote on post. Please try again: ${error}`);
    }
  };

  return (
    <div
      styleName={postId ? "post-excerpt" : "post-excerpt-hover"}
      data-testid="post"
    >
      <Votes
        voteStatus={props.data?.voteStatus ?? postData?.voteStatus}
        onVote={(
          vote: number,
          event: React.MouseEvent<HTMLImageElement, MouseEvent>
        ) => onVote(vote, event)}
        postId={postId ?? props.data?.id}
      />
      <div styleName="post-excerpt__content">
        <PostAuthor
          subredditName={props.data?.subredditName ?? postData?.subredditName}
          userName={props.data?.userName ?? postData?.userName}
          createdAt={props.data?.createdAt ?? postData?.createdAt}
        />
        <div styleName="post-excerpt__container">
          <h3 styleName="post-excerpt__title">
            {props.data?.title ?? postData?.title}
          </h3>
          <p
            styleName={`post-excerpt__description ${
              !postId && "post-excerpt__overflow"
            }`}
          >
            {props.data?.description ?? postData?.description}
          </p>
          <p styleName="post-excerpt__description">
            {props.data?.link ?? postData?.link}
          </p>
          {props.data?.imageURL ||
            (postData?.imageURL && (
              <div styleName="post-excerpt__image-container">
                <img
                  styleName="post-excerpt__image"
                  src={postData?.imageURL ?? props.data?.imageURL}
                  alt="post"
                />
              </div>
            ))}
        </div>
        <div styleName="post__buttons">
          <div styleName="post-excerpt__divider"></div>
          <PostInteractions
            commentsQuantity={
              props.data?.commentsQuantity ?? postData?.commentsQuantity
            }
            postId={postId}
          />
        </div>
      </div>
    </div>
  );
};

export default CSSModules(Post, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
