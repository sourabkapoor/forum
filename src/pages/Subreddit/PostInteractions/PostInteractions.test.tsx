import React from "react";
import { queryByText, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import PostInteractions from "./PostInteractions";
import Post from "../Post/Post";

describe("PostInteractions", () => {
  test("snapshot", () => {
    const { asFragment } = render(<PostInteractions commentsQuantity={0} postId="mock post id"/>);

    expect(asFragment()).toMatchSnapshot();
  });
  each([39, 251, 100, 512, 21, 51, 12, 1]).test("renders number of comments correctly", (a) => {
    render(<PostInteractions commentsQuantity={a} postId="mock post id"/>)

    const commentsQuantityText = screen.queryByText(`${a} Comments`)

    expect(commentsQuantityText).toBeInTheDocument();
  })
});
