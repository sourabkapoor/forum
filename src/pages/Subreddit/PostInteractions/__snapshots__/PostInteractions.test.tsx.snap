// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostInteractions snapshot 1`] = `
<DocumentFragment>
  <div
    class="post-excerpt__interactions"
  >
    <div
      class="post-excerpt__interaction"
    >
      <svg
        class="post-excerpt__icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 2H4c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h3v3.767L13.277 18H20c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zm0 14h-7.277L9 18.233V16H4V4h16v12z"
        />
      </svg>
      <span
        class="post-excerpt__action"
      >
        0 Comments
      </span>
    </div>
    <div
      class="post-excerpt__interaction"
    >
      <svg
        class="post-excerpt__icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M448 248L288 96v85.334C138.666 202.667 85.333 309.334 64 416c53.333-74.666 117.333-108.802 224-108.802v87.469L448 248z"
        />
      </svg>
      <span
        class="post-excerpt__action"
      >
        Share
      </span>
    </div>
    <div
      class="post-excerpt__interaction"
    >
      <svg
        class="post-excerpt__icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 384 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M336 0H48C21.49 0 0 21.49 0 48v464l192-112 192 112V48c0-26.51-21.49-48-48-48zm0 428.43l-144-84-144 84V54a6 6 0 0 1 6-6h276c3.314 0 6 2.683 6 5.996V428.43z"
        />
      </svg>
      <span
        class="post-excerpt__action"
      >
        Save
      </span>
    </div>
  </div>
</DocumentFragment>
`;
